library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -632,17 +693,29 @@ applyEquation term rule = fmap (either id Success) $ runExceptT $ do
                     pure $ substituteInTerm subst rule.rhs
                 unclearConditions -> throwE $ IndeterminateCondition unclearConditions
   where
-    -- evaluate/simplify a predicate, cut the operation short when it
-    -- is Bottom.
+    -- Simplify given predicate in a nested EquationT execution.
+    -- Return Nothing immediately if it is Bottom, return (Just
```

Comment:
Something that crossed my mind while looking at this code: 

According to the comment, `checkConstraint` returns `Just Nothing` if a constraint is true. If we end up with `[Just Nothing]`, than that would cause us to throw `IndeterminateCondition [Just Nothing]` in `applyEquation`.

However, the comment seems to be wrong: we will not return `Nothing` on false, but rather throw an exception, while `Nothing` will be returned on true and correctly filtered out by `catMaybes`.

Reply:
Good catch! I think this comment was explaining what a prior version without `whenBottom` was doing, which had a nest of `MaybeT ... (Maybe Predicate)` and got changed meanwhile. I adjusted the description.